### PR TITLE
Separate out the steps for creating a subscription

### LINKED ISSUE #235: Separate out an algorithm for creating a push subscription
This is currently only done when the developer calls `subscribe()`, but now that user agents can refresh subscriptions we'll have to reuse it.

### PR DESCRIPTION
**Pull request by beverloo**
Fixes #235. This builds on #253, the actual changes are included in the following commit:

https://github.com/w3c/push-api/pull/254/commits/56fa7fccf3059e60fa5b21fac4260bbb6e293139

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -383,6 +382,35 @@
...
+          To <dfn>create a push subscription</dfn>, given an <a>PushSubscriptionOptions</a> object
+          of <var>options</var>, the <a>user agent</a> must run the following steps:
+        </p>
+        <ol>
+          <li>Let <var>subscription</var> be a new <a>push subscription</a>.
+          </li>
+          <li>Set the <code>options</code> attribute of <var>subscription</var> to be a copy of
+          <var>options</var>.
+          </li>
+          <li>Generate a new P-256 <a>ECDH</a> key pair. Store the private key in an internal slot
```
Comment (martinthomson) [inline index.html]: Cite X9.62 on the first sentence.

File: index.html [hunk 2]
```diff
@@ -383,6 +382,35 @@
...
+          on <var>subscription</var>; this value MUST NOT be made available to applications. The
+          public key is also stored in an internal slot and can be retrieved by calling the <code>
+            getKey</code> method of the <a>PushSubscription</a> with an argument of
```
Comment (martinthomson) [inline index.html]: Is this indent what tidy produces?  Strange.

File: index.html [hunk 3]
```diff
@@ -383,6 +382,35 @@
...
+          </li>
+          <li>Generate a new authentication secret, which is a sequence of octets as defined in
+          [[!WEBPUSH-ENCRYPTION]]. Store the authentication secret in an internal slot on
+          <var>subscription</var>. This key can be retrieved by calling the <code>getKey</code>
+          method of the <a>PushSubscription</a> with an argument of <a data-link-for=
+          "PushEncryptionKeyName">auth</a>.
+          </li>
+          <li>Make a request to the <a>push service</a> to create a new <a>push subscription</a>.
+          Include the <a data-link-for="PushSubscriptionOptions">applicationServerKey</a> attribute
+          of <var>options</var> when it contains a valid point on the P-256 curve.
```
Comment (martinthomson) [inline index.html]: We reject the call if `applicationServerKey` is invalid, so you can remove the "when it contains a valid point" bit.

File: index.html [hunk 4]
```diff
@@ -868,9 +885,14 @@
         is one, or <code>null</code> otherwise.
       </p>
       <p>
+        When getting the <dfn>options</dfn> attribute, the <a>user agent</a> MUST return a
+        <code><a>PushSubscriptionOptions</a></code> object representing the options associated with
```
Comment (martinthomson) [inline index.html]: s/object/dictionary ?

File: index.html [hunk 5]
```diff
@@ -930,7 +953,7 @@
         </li>
       </ol>
       <p>
-        The serializer for a <a>PushSubscription</a> invokes the following steps:
+        The <dfn>serializer</dfn> for a <a>PushSubscription</a> invokes the following steps:
```
Comment (martinthomson) [inline index.html]: no data-dfn-for here?

**Review by martinthomson — APPROVED** [review]

File: index.html (hunk 2 above)
Comment (beverloo) [reply index.html]: Yes, presumably because the \<code> block is open?

File: index.html (hunk 3 above)
Comment (beverloo) [reply index.html]: "when it contains a valid point on the P-256 curve" -> "when it has been set"

File: index.html (hunk 4 above)
Comment (beverloo) [reply index.html]: No, because PSO is the object, PSOInit is the dictionary. We return PSO.

File: index.html (hunk 5 above)
Comment (beverloo) [reply index.html]: The owning \<section> has been annotated with the `data-dfn-for` attribute, making that redundant. That doesn't work when a section contains multiple identically named attributes, as is the case with the Init dictionaries.

File: index.html (hunk 1 above)
Comment (beverloo) [reply index.html]: I'm not sure how to phrase this. "as described in [[!X9.62]]"?

The actual curve parameters are defined in Section 10.2, should we refer that too?

Comment (beverloo) [reply index.html]: Went with just [[!X9.62]] since "as described in" doesn't add much either.

**Pull request merged**